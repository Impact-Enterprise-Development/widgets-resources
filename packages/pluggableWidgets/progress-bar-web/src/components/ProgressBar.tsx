import { createElement, CSSProperties, ReactElement, ReactNode } from "react";
import classNames from "classnames";
import { Alert } from "@widgets-resources/piw-utils";

export interface ProgressBarProps {
    class: string;
    style?: CSSProperties;
    currentValue: number;
    minValue: number;
    maxValue: number;
    onClick: (() => void) | undefined;
    label: ReactNode;
}

function calculatePercentage(currentValue: number, minValue: number, maxValue: number): number {
    if (currentValue < minValue) {
        return 0;
    }
    if (currentValue > maxValue) {
        return 100;
    }
    const range = maxValue - minValue;
    const percentage = Math.round(((currentValue - minValue) / range) * 100);
    return Math.abs(percentage);
}

function getValuesErrorMessage(currentValue: number, minValue: number, maxValue: number): string | null {
    if (maxValue < minValue) {
        return "Error in progress bar values: The maximum value is lower than the minimum value.";
    }
    if (currentValue < minValue) {
        return "Error in progress bar values: The progress value is lower than the minimum value.";
    }
    if (currentValue > maxValue) {
        return "Error in progress bar values: The progress value is higher than the maximum value.";
    }
    return null;
}

export function ProgressBar({
    class: className,
    style,
    minValue,
    maxValue,
    onClick,
    currentValue,
    label
}: ProgressBarProps): ReactElement {
    const errorMessage = getValuesErrorMessage(currentValue, minValue, maxValue);
    return (
        <div className={classNames("widget-progress-bar", "progress-bar-medium", className)} style={style}>
            <div
                className={classNames("progress", {
                    "widget-progress-bar-alert": maxValue < 1,
                    "widget-progress-bar-clickable": !!onClick
                })}
                onClick={onClick}
            >
                <div
                    className={classNames("progress-bar", "progress-bar-default", "progress-bar-medium", className)}
                    style={{ width: `${calculatePercentage(currentValue, minValue, maxValue)}%` }}
                >
                    {label}
                </div>
            </div>
            {errorMessage ? (
                <Alert bootstrapStyle="danger" className="widget-progress-bar">
                    {errorMessage}
                </Alert>
            ) : null}
        </div>
    );
}
