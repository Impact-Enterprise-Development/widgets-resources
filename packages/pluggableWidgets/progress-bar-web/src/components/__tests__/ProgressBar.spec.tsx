import { Alert } from "@widgets-resources/piw-utils";
import { mount, ReactWrapper, shallow } from "enzyme";
import { createElement, CSSProperties, FunctionComponent } from "react";
import { ProgressBar, ProgressBarProps } from "../ProgressBar";

const RandomComponent: FunctionComponent<any> = () => <div>This is a random component</div>;

describe("Progress bar", () => {
    function getProgressBarStyle(progressBar: ReactWrapper<ProgressBarProps>): CSSProperties | undefined {
        return progressBar.find(".progress-bar").prop("style");
    }
    const progress = 23;
    const maximumValue = 100;
    const onClickSpy = jest.fn();

    it("has progress bar structure", () => {
        const progressBar = shallow(
            <ProgressBar
                currentValue={progress}
                minValue={0}
                maxValue={maximumValue}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );
        expect(progressBar).toMatchSnapshot();
    });

    it("should show positive progress", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={progress}
                minValue={0}
                maxValue={maximumValue}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );

        expect(getProgressBarStyle(progressBar)).toEqual({ width: "23%" });
    });

    it("should call the onClick handler when the progress bar is clicked on", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={progress}
                minValue={0}
                maxValue={maximumValue}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );
        const progressElement = progressBar.find(".progress");
        expect(progressElement.hasClass("widget-progress-bar-clickable")).toBe(true);
        progressElement.simulate("click");
        expect(onClickSpy).toHaveBeenCalledTimes(1);
    });

    it("should properly handle a different range", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={progress}
                minValue={20}
                maxValue={30}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );

        // 23 on a range from 20 to 30 is 30% progress.
        expect(getProgressBarStyle(progressBar)).toEqual({ width: "30%" });
    });

    it("should clamp a current value lower than the minimum value to 0% progress", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={-20}
                minValue={0}
                maxValue={100}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );

        expect(getProgressBarStyle(progressBar)).toEqual({ width: "0%" });
    });

    it("should clamp a current value higher than the maximum value to 100% progress", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={110}
                minValue={0}
                maxValue={100}
                onClick={onClickSpy}
                label={`${progress}%`}
                class=""
            />
        );

        expect(getProgressBarStyle(progressBar)).toEqual({ width: "100%" });
    });

    it("should not be clickable when there is no onClick handler provided", () => {
        const progressBar = mount(
            <ProgressBar
                currentValue={50}
                minValue={0}
                maxValue={100}
                onClick={undefined}
                label={`${progress}%`}
                class=""
            />
        );
        expect(progressBar.find(".progress").hasClass("widget-progress-bar-clickable")).toBe(false);
    });

    describe("should show a runtime error Alert", () => {
        it("when the current value is lower than the minimum value", () => {
            const progressBar = mount(
                <ProgressBar
                    currentValue={-20}
                    minValue={0}
                    maxValue={100}
                    onClick={onClickSpy}
                    label={`${progress}%`}
                    class=""
                />
            );
            const alert = progressBar.find(Alert);
            expect(alert).toHaveLength(1);
            expect(alert.text()).toBe(
                "Error in progress bar values: The progress value is lower than the minimum value."
            );
        });

        it("when the current value is higher than the maximum value", () => {
            const progressBar = mount(
                <ProgressBar
                    currentValue={110}
                    minValue={0}
                    maxValue={100}
                    onClick={onClickSpy}
                    label={`${progress}%`}
                    class=""
                />
            );
            const alert = progressBar.find(Alert);
            expect(alert).toHaveLength(1);
            expect(alert.text()).toBe(
                "Error in progress bar values: The progress value is higher than the maximum value."
            );
        });

        it("when the range of the progress bar is negative", () => {
            const progressBar = mount(
                <ProgressBar
                    currentValue={50}
                    minValue={40}
                    maxValue={30}
                    onClick={onClickSpy}
                    label={`${progress}%`}
                    class=""
                />
            );
            const alert = progressBar.find(Alert);
            expect(alert).toHaveLength(1);
            expect(alert.text()).toBe(
                "Error in progress bar values: The maximum value is lower than the minimum value."
            );
        });
    });

    describe("the label of the progressbar", () => {
        it("should accept static text", () => {
            const progressBar = mount(
                <ProgressBar
                    currentValue={30}
                    minValue={0}
                    maxValue={100}
                    onClick={onClickSpy}
                    label="This is your progress"
                    class=""
                />
            );
            expect(progressBar.text()).toBe("This is your progress");
        });

        it("should accept a component", () => {
            const progressBar = mount(
                <ProgressBar
                    currentValue={30}
                    minValue={0}
                    maxValue={100}
                    onClick={onClickSpy}
                    label={<RandomComponent />}
                    class=""
                />
            );
            expect(progressBar.find(RandomComponent)).toHaveLength(1);
            expect(progressBar.text()).toContain("This is a random component");
        });

        it("should accept nothing", () => {
            const progressBar = mount(
                <ProgressBar currentValue={30} minValue={0} maxValue={100} onClick={onClickSpy} label={null} class="" />
            );
            expect(progressBar.text()).toHaveLength(0);
        });
    });
});
