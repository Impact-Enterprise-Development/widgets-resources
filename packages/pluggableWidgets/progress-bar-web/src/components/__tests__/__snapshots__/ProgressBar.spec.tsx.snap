// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Progress bar has progress bar structure 1`] = `
<div
  className="widget-progress-bar progress-bar-medium"
>
  <div
    className="progress widget-progress-bar-clickable"
    onClick={[MockFunction]}
  >
    <div
      className="progress-bar progress-bar-default"
      style={
        Object {
          "width": "23%",
        }
      }
    >
      23%
    </div>
  </div>
</div>
`;
